lib/Platform/Intl/BCP47Parser.cpp

```diff
@@ -0,0 +1,539 @@
+  ~LanguageTagParser();
+
+  // public function declaration
+  bool parseUnicodeLocaleId();
+  ParsedLocaleIdentifier getParsedLocaleId();
+  std::u16string toString();
+
+ private:
+  // private function declaration
+  bool parseUnicodeLanguageId(bool transformedExtensionId);
+  bool addVariantSubtag(bool transformedExtensionId);
+  bool parseExtensions();
+  bool parseUnicodeExtension();
+  bool parseTransformedExtension();
+  bool parseOtherExtension(char16_t singleton);
+  bool parsePUExtension();
+  // tokenizer functions
+  std::u16string getCurrentSubtag();
+  llvh::ArrayRef<char16_t> getCurrentSubtagRef();
+  bool hasMoreSubtags();
+  bool nextSubtag();
+  
+  ParsedLocaleIdentifier parsedLocaleIdentifier;
+  std::u16string localeId_;
+  const std::u16string &originalLocaleId_;
+  int subtagStart_;
+  int subtagEnd_;
+};
+
+LanguageTagParser::LanguageTagParser(const std::u16string &localeId) : localeId_{localeId}, originalLocaleId_(localeId), subtagStart_(0), subtagEnd_(-1) {
+  toASCIILowerCase(localeId_);
+}
+LanguageTagParser::~LanguageTagParser() = default;
+
+ParsedLocaleIdentifier LanguageTagParser::getParsedLocaleId(){
+  return parsedLocaleIdentifier;
+}
+
+bool LanguageTagParser::parseUnicodeLocaleId() {
+  if (!parseUnicodeLanguageId(false))  {
+    return false;
+  }
+  if (!hasMoreSubtags()) {
+    return true;
+  }
+  if (!parseExtensions()) {
+    return false;
+  }
+  
+  return !hasMoreSubtags();
+}
+
+bool LanguageTagParser::parseUnicodeLanguageId(bool transformedExtensionId) {
+  if (!transformedExtensionId) {
+    if (!nextSubtag() || !isUnicodeLanguageSubtag(getCurrentSubtagRef())) {
+      return false;
+    }
+  }
+  
+  ParsedLanguageIdentifier &languageId = transformedExtensionId ? parsedLocaleIdentifier.transformedLanguageIdentifier : parsedLocaleIdentifier.languageIdentifier;
+  
+  languageId.languageSubtag = getCurrentSubtag();
+  
+  if (!nextSubtag()) {
+    return true;
+  }
+  
+  if (isUnicodeScriptSubtag(getCurrentSubtagRef())) {
+    languageId.scriptSubtag = getCurrentSubtag();
+    if (!transformedExtensionId) {
+      toASCIITitleCase(languageId.scriptSubtag);
+    }
+    if (!nextSubtag()) {
+      return true;
+    }
+  }
+
+  if (isUnicodeRegionSubtag(getCurrentSubtagRef())) {
+    languageId.regionSubtag = getCurrentSubtag();
+    if (!transformedExtensionId) {
+      toASCIIUpperCase(languageId.regionSubtag);
+    }
+    if (!nextSubtag()) {
+      return true;
+    }
+  }
+  
+  while (true) {
+    if (!isUnicodeVariantSubtag(getCurrentSubtagRef())) {
+      return true;
+    } else if (!addVariantSubtag(transformedExtensionId)) {
+      return false;
+    }
+  
+    if (!nextSubtag()) {
+      return true;
+    }
+  }
+}
+
+bool LanguageTagParser::addVariantSubtag(bool transformedExtensionId) {
+  ParsedLanguageIdentifier &languageId = transformedExtensionId ? parsedLocaleIdentifier.transformedLanguageIdentifier : parsedLocaleIdentifier.languageIdentifier;
+
+  // Insert in alphabetical order
```

Comment:
We should either sort them later, or use a `std::set`. If we can fit sorting in a logical way, then I have a slight preference for that, otherwise either is fine.